src/nccl_ofi_idpool.cpp

```diff
@@ -4,116 +4,53 @@
 
 #include "config.h"
 
-#include <assert.h>
-#include <errno.h>
-#include <stdlib.h>
+#include <algorithm>
+#include <mutex>
+#include <vector>
 
 #include "nccl_ofi_idpool.h"
 #include "nccl_ofi_math.h"
-#include "nccl_ofi_pthread.h"
+#include "nccl_ofi_log.h"
 
-/*
- * @brief	Initialize pool of IDs
- *
- * Allocates and initializes a nccl_ofi_idpool_t object, marking all
- * IDs as available.
- *
- * @param	idpool_p
- *		Return value with the ID pool pointer allocated
- * @param	size
- *		Size of the id pool (number of IDs)
- * @return	0 on success
- *		non-zero on error
- */
-int nccl_ofi_idpool_init(nccl_ofi_idpool_t *idpool, size_t size)
-{
-	int ret = 0;
-
-	assert(NULL != idpool);
-
-	if (0 == size) {
-		/* Empty or unused pool */
-		idpool->ids = NULL;
-		idpool->size = 0;
-		return ret;
-	}
 
-	/* Scale pool size to number of 64-bit uints (rounded up) */
-	size_t num_long_elements = NCCL_OFI_ROUND_UP(size, sizeof(uint64_t) * 8) / (sizeof(uint64_t) * 8);
-
-	/* Allocate memory for the pool */
-	idpool->ids = (uint64_t *)malloc(sizeof(uint64_t) * num_long_elements);
-
-	/* Return in case of allocation error */
-	if (NULL == idpool->ids) {
-		NCCL_OFI_WARN("Unable to allocate ID pool");
-		return -ENOMEM;
+nccl_ofi_idpool_t::nccl_ofi_idpool_t(const size_t& size) : 
+	size_(size)
+{
+	if (size_ == 0) {
+		idpool_ = std::vector<uint64_t>();
+		return;
 	}
 
-	/* Set all IDs to be available */
-	memset(idpool->ids, 0xff, size / 8);
-	if (size % 8) {
-		idpool->ids[num_long_elements - 1] = (1ULL << (size % (sizeof(uint64_t) * 8))) - 1;
-	}
+	size_t num_long_elements = NCCL_OFI_ROUND_UP(size_, sizeof(uint64_t) * 8) / (sizeof(uint64_t) * 8);
+	idpool_ = std::vector<uint64_t>(num_long_elements, 0xffffffffffffffff);
 
-	/* Initialize mutex */
-	ret = nccl_net_ofi_mutex_init(&idpool->lock, NULL);
-	if (OFI_UNLIKELY(ret)) {
-		NCCL_OFI_WARN("Unable to initialize mutex");
-		free(idpool->ids);
-		idpool->ids = NULL;
-		return ret;
+	if (size_ % 8 != 0) {
```

Comment:
throw a comment in here about what this is doing.

Reply:
Added in force push, although my comment may need a revision for clarity.